cEP-0018.md

```diff
@@ -0,0 +1,199 @@
+
+Managing a new repository is a heavy task, and this will be highly automated as
+follows:
+
+1. The parsers in coala-antlr/coantparsers would be generated and
+   pushed via travis-builds whenever a new commit is pushed.
+2. The cib tool can be enhanced to deal with the installation of bears that
+   require only some specified parsers (for e.g a PyUnusedVarBear would
+   only require parser for python).
+3. The cib tool can also trigger specialised builds and download the newly
+   generated parser on the fly.
+
+## Code Samples/Prototypes
+
+Here is a prototype for the implementations within coantlib:
+
+```python
+import antlr4
+import coantparsers
+
+from coalib.bearlib.languages.definitions import *
+from coalib.bearlib.languages.Language import parse_lang_str, Languages
+
+class ASTLoader():
+    mapping = {
+        C : [coantparsers.clexer, coantparsers.cparser],
+        Python : [coantparsers.pylexer, coantparsers.pyparser],
+        JavaScript : [coantparsers.jslexer, coantparsers.jsparser],
+        ...
+    }
+
+    @staticmethod
+    def loadFile(file,filename,lang = 'auto'):
+        """
+        Loads file's AST into memory
+        """
+        if(lang == 'auto'):
+          ext = get_extension(filename)
+        else:
+          ext = Languages(parse_lang_str(lang)[0])[0]
+        if ext is None:
+            raise FileExtensionNotSupported('Unknown File Type')
+        inputStream = file.readlines()
+        lexer = mapping[ext][0](inputStream)
+        parser = mapping[ext][1](lexer)
+        return parser
+
+class ASTWalker():
+    treeRoot = None
+    tree = None
+    def __init__(file,filename):
+        self.tree = ASTLoader.load(file, filename)
+        self.treeRoot = self.tree
+
+    def get_next_node(self):
+        """
+        Modify the internal tree variable and return the next node
+        """
+
+    def get_prev_node(self):
+        """
+        Modify the internal tree variable and return prev node
+        """
+
+    def reset(self):
+        """
+        Reset the internal tree variable to tree root
+        """
+
+    def is_terminal_node(node):
+        """
+        Determine if the node is terminal
+        """
+        return isinstance(node, antlr4.tree.Tree.TerminalNode)
+
+    def get_imports():
+        """
+        Return list of strings containing import statements from the file
+        Requires the LANGUAGE parameter to contain exactly one language
+        """
+
+    def get_methods():
+        """
+        Return list of strings containing all methods from the file
+        Requires the LANGUAGE parameter to contain exactly one language
+        """
+    ...
+```
+
+### Prototype of  `ASTBear` class implementation
+
+```python
+from coantlib import ASTWalker
+from coalib.bears.LocalBear import Bear
+
+class ASTBear(LocalBear):
+    walker = None
+
+    def initialise(file, filename):
+        if LANGUAGES:
+            self.walker = ASTWalker(file, filename, self.LANGUAGES[0])
```

Comment:
`ASTWalker(file, filename, self.LANGUAGES[0])` should return PythonASTWalker if `self.LANGUAGES == 'python'`